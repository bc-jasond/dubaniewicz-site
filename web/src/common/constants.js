export const PAGE_NAME_VIEW = 'public';
export const PAGE_NAME_EDIT = 'edit';
export const PAGE_NAME_PUBLIC = 'public';
export const PAGE_NAME_PRIVATE = 'private';
export const PAGE_NAME_MANAGE = 'manage';
export const PAGE_NAME_USER_PROFILE = 'user';

// EDITOR
export const NEW_POST_URL_ID = 'new';
// TODO: this is a placeholder to be able to set the caret in an empty tag
export const ZERO_LENGTH_CHAR = '\u200B';
// key codes from a 2015 macbook 13" & Kinesis Advantage2
// TODO: keyCode is deprecated
export const KEYCODE_BACKSPACE = 8;
export const KEYCODE_TAB = 9;
export const KEYCODE_ENTER = 13;
export const KEYCODE_SHIFT_RIGHT = 16;
export const KEYCODE_CTRL = 17;
export const KEYCODE_ALT = 18;
export const KEYCODE_CAPS_LOCK = 20;
export const KEYCODE_ESC = 27;
export const KEYCODE_SPACE = 32;
export const KEYCODE_PAGE_UP = 33;
export const KEYCODE_PAGE_DOWN = 34;
export const KEYCODE_END = 35;
export const KEYCODE_HOME = 36;
export const KEYCODE_LEFT_ARROW = 37;
export const KEYCODE_UP_ARROW = 38;
export const KEYCODE_RIGHT_ARROW = 39;
export const KEYCODE_DOWN_ARROW = 40;
export const KEYCODE_DEL = 46;
export const KEYCODE_V = 86;
export const KEYCODE_X = 88;
export const KEYCODE_Z = 90;
export const KEYCODE_Y = 89;
export const KEYCODE_SHIFT_OR_COMMAND_LEFT = 91;
export const KEYCODE_COMMAND_RIGHT = 93;
export const KEYCODE_F1 = 112;
export const KEYCODE_F2 = 113;
export const KEYCODE_F3 = 114;
export const KEYCODE_F4 = 115;
export const KEYCODE_F5 = 116;
export const KEYCODE_F6 = 117;
export const KEYCODE_F7 = 118;
export const KEYCODE_F8 = 119;
export const KEYCODE_F9 = 120;
export const KEYCODE_F10 = 121;
export const KEYCODE_F11 = 122;
export const KEYCODE_F12 = 123;
export const KEYCODE_PRINT_SCREEN = 124;
export const KEYCODE_SPACE_NBSP = 160;

// post level edits
export const POST_ACTION_REDIRECT_TIMEOUT = 1000;

// ENV
export const API_URL =
  process.env.NODE_ENV === 'production'
    ? 'https://api.filbert.xyz'
    : 'http://localhost:3001';

// AUTH
export const FILBERT_LOCALSTORAGE_NAMESPACE = 'filbert';
export const AUTH_TOKEN_KEY = 'token';
export const SESSION_KEY = 'session';
export const SESSION_THEME = 'theme';
export const SESSION_FONT = 'font';

// DOM
export const DOM_ELEMENT_NODE_TYPE_ID = 1;
export const DOM_TEXT_NODE_TYPE_ID = 3;
export const DOM_INPUT_TAG_NAME = 'INPUT';


// THEME
export const DARK_MODE_THEME = 'dark';
export const LIGHT_MODE_THEME = 'light';
export const SANS_FONT_THEME = 'sans';
export const MIXED_FONT_THEME = 'mixed';
