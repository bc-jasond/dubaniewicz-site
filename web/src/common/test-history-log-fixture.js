export const nodeUpdatesByNodeId = [
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some co',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context h',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://www.',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://www.stuff.c',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://www.stuff.com',
    },
    id: '0fbe',
  },
  {
    post_id: 277,
    type: 'quote',
    content: '',
    meta: {
      quote: "Here's something to say",
      author: 'B Guys and Guy',
      context: 'some context here',
      url: 'http://www.stuff.com',
    },
    id: '0fbe',
    next_sibling_id: 'a43f',
  },
  { post_id: 277, type: 'p', content: '', meta: {}, id: 'a43f' },
  { post_id: 277, type: 'p', content: 'And ano', meta: {}, id: 'a43f' },
  { post_id: 277, type: 'p', content: 'And another?', meta: {}, id: 'a43f' },
  {
    post_id: 277,
    type: 'p',
    content: 'And another?',
    meta: {},
    id: 'a43f',
    next_sibling_id: '1744',
  },
  { post_id: 277, type: 'p', content: '', meta: {}, id: '1744' },
  {
    post_id: 277,
    type: 'p',
    content: 'And another?',
    meta: {
      selections: {
        length: -1,
        'selection-bold': false,
        'selection-italic': false,
        'selection-code': false,
        'selection-siteinfo': false,
        'selection-mini': false,
        'selection-strikethrough': false,
        'selection-link': false,
        linkUrl: '',
      },
    },
    id: 'a43f',
    next_sibling_id: '1744',
  },
  '1744',
];
export const contentNodeHistoryLog = [
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some co',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some co',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context h',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context h',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.c',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '0fbe' },
    unexecuteSelectionOffsets: { startNodeId: '0fbe' },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.c',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.com',
          },
          id: '0fbe',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: 'a43f', caretStart: 0 },
    unexecuteSelectionOffsets: {
      startNodeId: '0fbe',
      caretStart: 0,
      caretEnd: 0,
    },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.com',
          },
          id: '0fbe',
        },
        executeState: {
          post_id: 277,
          type: 'quote',
          content: '',
          meta: {
            quote: "Here's something to say",
            author: 'B Guys and Guy',
            context: 'some context here',
            url: 'http://www.stuff.com',
          },
          id: '0fbe',
          next_sibling_id: 'a43f',
        },
      },
      {
        executeState: {
          post_id: 277,
          type: 'p',
          content: '',
          meta: {},
          id: 'a43f',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: 'a43f', caretStart: 7 },
    unexecuteSelectionOffsets: {
      caretStart: 0,
      caretEnd: 0,
      startNodeId: 'a43f',
    },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: '',
          meta: {},
          id: 'a43f',
        },
        executeState: {
          post_id: 277,
          type: 'p',
          content: 'And ano',
          meta: {},
          id: 'a43f',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: 'a43f', caretStart: 12 },
    unexecuteSelectionOffsets: {
      caretStart: 7,
      caretEnd: 7,
      startNodeId: 'a43f',
    },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: 'And ano',
          meta: {},
          id: 'a43f',
        },
        executeState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {},
          id: 'a43f',
        },
      },
    ],
  },
  {
    executeSelectionOffsets: { startNodeId: '1744', caretStart: 0 },
    unexecuteSelectionOffsets: {
      caretStart: 12,
      caretEnd: 12,
      startNodeId: 'a43f',
    },
    historyState: [
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {},
          id: 'a43f',
        },
        executeState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {},
          id: 'a43f',
          next_sibling_id: '1744',
        },
      },
      {
        executeState: {
          post_id: 277,
          type: 'p',
          content: '',
          meta: {},
          id: '1744',
        },
      },
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {},
          id: 'a43f',
          next_sibling_id: '1744',
        },
        executeState: {
          post_id: 277,
          type: 'p',
          content: 'And another?',
          meta: {
            selections: {
              length: -1,
              'selection-bold': false,
              'selection-italic': false,
              'selection-code': false,
              'selection-siteinfo': false,
              'selection-mini': false,
              'selection-strikethrough': false,
              'selection-link': false,
              linkUrl: '',
            },
          },
          id: 'a43f',
          next_sibling_id: '1744',
        },
      },
      {
        unexecuteState: {
          post_id: 277,
          type: 'p',
          content: '',
          meta: {},
          id: '1744',
        },
      },
    ],
  },
];
