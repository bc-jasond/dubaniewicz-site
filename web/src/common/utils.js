import { Map } from 'immutable';
import { s4 } from '@filbert/util';

export function confirmPromise(msg) {
  return new Promise((resolve) => {
    if (confirm(msg)) {
      resolve(true);
      return;
    }
    resolve(false);
  });
}

export function formatPostDate(dateStr) {
  if (!dateStr) {
    return;
  }
  const publishedDate = new Date(dateStr);
  return publishedDate.toLocaleDateString('en-us', {
    // weekday: 'long',
    year: 'numeric',
    day: 'numeric',
    month: 'long',
  });
}

export function formatStreakDate(dateInt) {
  const date = new Date();
  date.setFullYear(parseInt(`${dateInt}`.substring(0, 4), 10));
  date.setMonth(parseInt(`${dateInt}`.substring(4, 6), 10));
  date.setDate(parseInt(`${dateInt}`.substring(6), 10));
  return date.toLocaleDateString('en-us', {
    // weekday: 'long',
    year: 'numeric',
    day: 'numeric',
    month: 'long',
  });
}

export function formatNumber(number) {
  return new Intl.NumberFormat().format(number);
}

export function getCharFromEvent(evt) {
  // Firefox issue when typing fast - skip these.  NOTE: this could ignore composing characters for languages other than English
  //https://stackoverflow.com/a/25509350/1991322
  if (evt.key === 'Process') {
    return '';
  }
  if (evt && typeof evt.keyCode !== 'undefined') {
    // for normal "found on the keyboard" characters
    return evt.key;
  }
  // for OS X emoji keyboard insert
  return evt.data;
}

export function getCanonicalFromTitle(title) {
  let canonical = title
    // take first 25 chars
    .substring(0, 25)
    .toLowerCase()
    // remove all whitespace, replace with hyphen
    .replace(/\s+/g, '-')
    // keep only alpha numeric chars
    .replace(/[^0-9a-z-]/g, '');
  // append a hash
  canonical += `-${s4()}`;
  // dedupe hyphens
  return canonical.replace(/-+/g, '-');
}

export function stopAndPrevent(evt) {
  if (evt && evt.stopPropagation && evt.preventDefault) {
    evt.stopPropagation();
    evt.preventDefault();
  }
}

export function usernameIsValid(maybeUsername) {
  return /^@[0-9a-z]{1,42}$/.test(maybeUsername);
}

export function idIsValid(maybeId) {
  return new RegExp(/[0-9a-f]{4}/).test(maybeId);
}

export function nodeIsValid(node) {
  return Map.isMap(node) && idIsValid(node.get('id'));
}

export function isBrowser() {
  return typeof window !== 'undefined';
}

export function getUrl(s) {
  try {
    const url = new URL(s);
    const urlString = url.toString();
    // URL().toString() adds the trailing / slash to the host portion of the url
    // this is annoying when you're typing
    // do a check to see if the end of the user input is a `/` and if not, strip it
    if (urlString.slice(-1) === '/' && s.slice(-1) !== '/') {
      return urlString.slice(0, -1);
    }
    return urlString;
  } catch (err) {
    return '';
  }
}
