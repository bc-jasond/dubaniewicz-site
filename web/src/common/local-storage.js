import { fromJS, isImmutable, Map } from 'immutable';
import { FILBERT_LOCALSTORAGE_NAMESPACE } from './constants';

// reading & parsing JSON from localStorage is too slow.
// take a dual-write approach - one sync to memory and one optimistic async to localStorage
let inMemoryCache = Map();
let writeTimeout = {};

// TODO: this is a cheap way out, this file should be mocked in consumer tests
export function clearForTests() {
  inMemoryCache = Map();
  writeTimeout = {};
}

function getNamespacedKey(key) {
  return `${FILBERT_LOCALSTORAGE_NAMESPACE}-${key}`;
}

export function set(key, value, debounce = true) {
  if (!value) {
    inMemoryCache = inMemoryCache.delete(key);
  } else {
    inMemoryCache = inMemoryCache.set(key, fromJS(value));
  }

  if (writeTimeout[key]) {
    clearTimeout(writeTimeout[key]);
  }
  const setLocalStorage = (k) => {
    try {
      const namespacedKey = getNamespacedKey(k);
      let item = inMemoryCache.get(k);
      if (isImmutable(item)) {
        item = item.toJS();
      }
      if (!item) {
        localStorage.removeItem(namespacedKey);
      } else {
        localStorage.setItem(namespacedKey, JSON.stringify(item));
      }
    } catch (err) {
      // sometimes the localStorage.setItem() call fails - probably exceeded quota
      console.error(err);
      if (err?.name === 'QuotaExceededError') {
        // truncate history
      }
    }
  };
  if (!debounce) {
    setLocalStorage(key);
    return;
  }
  // this is a "backup" of the in memory storage to localStorage (for page reload in the same browser)
  // TODO: this could be written to a central storage like the DB or redis
  // this is a slow operation, only do it every so often
  writeTimeout[key] = setTimeout(setLocalStorage.bind(null, key), 3000);
}

export function get(key, defaultValue) {
  const namespacedKey = getNamespacedKey(key);
  try {
    if (!inMemoryCache.has(key)) {
      const currentLocal = localStorage.getItem(namespacedKey);
      if (
        currentLocal &&
        currentLocal !== 'undefined' &&
        currentLocal !== 'null'
      ) {
        const jsValue = JSON.parse(currentLocal);
        const woke = fromJS(jsValue);
        inMemoryCache = inMemoryCache.set(key, woke);
      }
    }
  } catch (err) {
    console.error(err);
    // if JSON.parse() fails, just blow this key away!
    set(key, undefined, false);
  }
  return inMemoryCache.get(key) || defaultValue;
}
