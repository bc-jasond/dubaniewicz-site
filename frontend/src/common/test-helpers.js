import { fromJS } from 'immutable';
import { reviver } from './utils';
import { SELECTION_LENGTH, SELECTION_NEXT } from './constants';

export function overrideConsole() {
  global.console = {
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn()
  };
}

export function mockLocalStorage() {
  global.localStorageStorage = {};
  global.localStorage = {
    getItem: k => global.localStorageStorage[k],
    setItem: (k, v) => (global.localStorageStorage[k] = v),
    clear: () => (global.localStorageStorage = {})
  };
}

export function mockSetTimeout() {
  global.setTimeout = jest.fn();
}

export function makeSelections(values) {
  const head = {};
  let current = head;
  let prev;
  do {
    let [currentLength, ...currentValues] = values.shift();
    current[SELECTION_LENGTH] = currentLength;
    currentValues.forEach(v => {
      if (typeof v === 'object') {
        current[v.key] = v.value;
      } else {
        current[v] = true;
      }
    });
    if (prev) {
      prev[SELECTION_NEXT] = current;
    }
    prev = current;
    current = {};
  } while (values.length);
  prev[SELECTION_NEXT] = undefined; //important for the reviver()
  prev[SELECTION_LENGTH] = -1;
  return fromJS(head, reviver);
}
