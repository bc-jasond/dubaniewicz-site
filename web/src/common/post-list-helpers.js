import { getApiClientInstance } from './api-client';
import { formatPostDate } from './utils';
import { loading } from '../stores';

export function pushHistory(param, value) {
  const urlSearchParams = new URLSearchParams(window.location.search);
  if (value) {
    urlSearchParams.set(param, value === true ? '' : value);
  } else {
    urlSearchParams.delete(param);
  }
  const updatedQueryString =
    urlSearchParams.toString().length > 0
      ? `?${urlSearchParams.toString()}`
      : '';
  // update the URL in history for the user to retain, only if it changed!
  if (window.location.search != updatedQueryString) {
    window.history.pushState(
      {},
      document.title,
      `${window.location.pathname}${updatedQueryString}`
    );
  }
  return urlSearchParams;
}

export async function loadPosts(url, urlSearchParams, fetchClient) {
  loading.set(true);
  const apiClient = getApiClientInstance(fetchClient);
  const params = urlSearchParams.toString();
  const queryString = params.length > 0 ? `?${params}` : '';
  const { error, data: postsData } = await apiClient.get(
    `${url}${queryString}`
  );
  loading.set(false);
  return postsData.map((post) => ({
    ...post,
    published: formatPostDate(post.published),
    updated: formatPostDate(post.updated),
  }));
}
