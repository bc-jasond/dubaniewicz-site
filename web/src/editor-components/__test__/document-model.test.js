import { Map } from 'immutable';
import { overrideConsole } from '../../common/test-helpers';

import * as SelectionHelpers from '../selection-helpers';

import DocumentModel, { getFirstNode, getLastNode } from '../document-model';
import {
  firstNodeIdH1,
  h2Id,
  testPostWithAllTypesJS,
} from '../../common/test-post-with-all-types';

overrideConsole();

const { post, contentNodes } = testPostWithAllTypesJS;

let documentModel;

beforeEach(() => {
  documentModel = DocumentModel(post.id, contentNodes);
});

describe('DocumentModel', () => {
  test('getNode', () => {
    expect(documentModel.getNode(null)).toBe(Map());
    // finds: root -> content -> ol -> list item
    expect(documentModel.getNode(firstNodeIdH1)).toMatchSnapshot();
    expect(documentModel.getNode('badId')).toBe(Map());
  });
  test('getNextNode', () => {
    // returns Map() when node is at last position
    expect(documentModel.getNextNode('cce3')).toBe(Map());
    expect(documentModel.getNextNode('7a38')).toMatchSnapshot();
  });
  test('getPrevNode', () => {
    // returns Map() when node is at position 0
    expect(documentModel.getPrevNode(firstNodeIdH1)).toBe(Map());
    // nominal case
    expect(documentModel.getPrevNode('4add')).toMatchSnapshot();
  });
  test('isFirstOfType', () => {
    // positive case
    expect(documentModel.isFirstOfType('56da')).toBe(true);
    // negative case - last child
    expect(documentModel.isFirstOfType('9fa0')).toBe(false);
  });
  test('isLastOfType', () => {
    // positive case
    expect(documentModel.isLastOfType('cce3')).toBe(true);
    // negative case - last child
    expect(documentModel.isLastOfType('4b8c')).toBe(false);
  });
  test.todo('getNodesBetween');
  test.todo('getFirstNode');
  test.todo('getLastNode');
  test.todo('insert');
  test('isTextType', () => {
    [
      // P
      'cce3',
      // LI
      '151c',
      // H1
      firstNodeIdH1,
      // h2
      h2Id,
      // Pre
      'fd25',
    ].forEach((nodeId) => {
      expect(documentModel.isTextType(nodeId)).toBe(true);
      expect(documentModel.isMetaType(nodeId)).toBe(false);
    });
    [
      // SPACER
      'db70',
      // IMAGE
      '4add',
      // QUOTE
      'c67c',
    ].forEach((nodeId) => {
      expect(documentModel.isTextType(nodeId)).toBe(false);
      expect(documentModel.isMetaType(nodeId)).toBe(true);
    });
  });
  test('mergeParagraphs', () => {
    const combinedContent = `${documentModel
      .getNode('621e')
      .get('content')}${documentModel.getNode('f677').get('content')}`;
    const spy = jest
      .spyOn(SelectionHelpers, 'concatSelections')
      .mockImplementation((arg) => arg);
    documentModel.mergeParagraphs('621e', 'f677');
    expect(spy).toHaveBeenCalled();
    expect(documentModel.getNode('621e').get('content')).toBe(combinedContent);
  });
  //test.todo('update() - this is tested in history-manager.test.js');
  //test.todo('delete() - this is tested in history-manager.test.js');
});
