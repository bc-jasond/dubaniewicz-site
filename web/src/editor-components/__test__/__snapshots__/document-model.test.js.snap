// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentModel getNextNode 1`] = `
Immutable.Map {
  "post_id": 175,
  "id": "43eb",
  "next_sibling_id": "4add",
  "type": "pre",
  "content": "getIt();",
  "meta": Immutable.Map {},
}
`;

exports[`DocumentModel getNode 1`] = `
Immutable.Map {
  "post_id": 175,
  "id": "8e34",
  "next_sibling_id": "621e",
  "type": "h1",
  "content": "Large Heading",
  "meta": Immutable.Map {},
}
`;

exports[`DocumentModel getPrevNode 1`] = `
Immutable.Map {
  "post_id": 175,
  "id": "43eb",
  "next_sibling_id": "4add",
  "type": "pre",
  "content": "getIt();",
  "meta": Immutable.Map {},
}
`;
